File: 2025-12/src/day_20/boj_1300/Main.cpp
#include <bits/stdc++.h>
using namespace std;

long long lower_bound_param(long long n, int k) {
    long long left = 1;
    long long right = min(1000000000LL, n * n);

    while (left < right) {
        long long mid = (left + right) / 2;

        long long cnt = 0;
        for (int i = 1; i <= n; i++) {
            cnt += min(mid / i, n);
        }

        if (cnt < k) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }

    return right;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, k;
    cin >> n >> k;
    cout << lower_bound_param(n, k);
}

File: 2025-12/src/day_20/boj_1300/Main.java
package day_20.boj_1300;

import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        int k = Integer.parseInt(br.readLine());

        System.out.println(lowerBound(n, k));
    }

    static int lowerBound(int n, int k) {
        int left = 1;
        int right = (int) Math.min(1_000_000_000, (long) n * n);

        while (left < right) {
            int mid = (left + right) / 2;

            long cnt = 0;
            for (int i = 1; i <= n; i++) {
                cnt += Math.min(mid / i, n);
            }

            if (cnt < k) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }

        return right;
    }
}
